python/ray/tests/test_placement_group_5.py

```diff
@@ -450,6 +450,118 @@ def verify_actor_killed():
+    cluster.add_node(num_cpus=8, resources={"head": 1})
+    cluster.wait_for_nodes()
+    ray.init(address=cluster.address)
+    cluster.add_node(num_cpus=2, resources={"worker1": 1})
+    worker2_node = cluster.add_node(num_cpus=4, resources={"worker2": 1})
+    cluster.wait_for_nodes()
+
+    @ray.remote
+    def get_node_id():
+        return ray.get_runtime_context().get_node_id()
+
+    head_node_id = ray.get(get_node_id.options(resources={"head": 1}).remote())
+    worker1_node_id = ray.get(get_node_id.options(resources={"worker1": 1}).remote())
+    worker2_node_id = ray.get(get_node_id.options(resources={"worker2": 1}).remote())
+
+    # soft_target_node_id only works with STRICT_PACK
+    with pytest.raises(ValueError):
+        pg = ray.util.placement_group(
+            bundles=[{"CPU": 2}, {"CPU": 2}],
+            strategy="PACK",
+            _soft_target_node_id=ray.NodeID.from_random().hex(),
+        )
+
+    # Invalid target node id
+    with pytest.raises(ValueError):
+        pg = ray.util.placement_group(
+            bundles=[{"CPU": 2}, {"CPU": 2}], strategy="PACK", _soft_target_node_id="a"
+        )
+
+    # No target node.
+    pg = ray.util.placement_group(
+        bundles=[{"CPU": 2}, {"CPU": 2}], strategy="STRICT_PACK"
+    )
+    wait_for_condition(lambda: ray.available_resources()["CPU"] == 10)
+    assert (
+        ray.get(
+            get_node_id.options(
+                scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=pg)
+            ).remote()
+        )
+        == head_node_id
+    )
+    ray.util.remove_placement_group(pg)
+    wait_for_condition(lambda: ray.available_resources()["CPU"] == 14)
+
+    # Target node doesn't have enough available resources.
+    pg = ray.util.placement_group(
+        bundles=[{"CPU": 2}, {"CPU": 2}],
+        strategy="STRICT_PACK",
+        _soft_target_node_id=worker1_node_id,
+    )
+    wait_for_condition(lambda: ray.available_resources()["CPU"] == 10)
+    assert (
+        ray.get(
+            get_node_id.options(
+                scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=pg)
+            ).remote()
+        )
+        == head_node_id
+    )
+    ray.util.remove_placement_group(pg)
+    wait_for_condition(lambda: ray.available_resources()["CPU"] == 14)
+
+    # Target node doesn't exist.
+    pg = ray.util.placement_group(
+        bundles=[{"CPU": 2}, {"CPU": 2}],
+        strategy="STRICT_PACK",
+        _soft_target_node_id=ray.NodeID.from_random().hex(),
+    )
+    wait_for_condition(lambda: ray.available_resources()["CPU"] == 10)
+    assert (
+        ray.get(
+            get_node_id.options(
+                scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=pg)
+            ).remote()
+        )
+        == head_node_id
+    )
+    ray.util.remove_placement_group(pg)
+    wait_for_condition(lambda: ray.available_resources()["CPU"] == 14)
+
+    # Target node has enough available resources.
```

Comment:
iterate multiple times to make sure it is not a lucky case? (e.g., )

Reply:
I set up the test in a way that this target node is not the preferred node if `_soft_target_node_id` is not set.

Also we run it for every commit, if it's lucky case, it will show as flaky eventually.